docs/versioning.md

```diff
@@ -0,0 +1,114 @@
+example, the third attempt at fulfilling the ASM release corresponding to Istio
+1.7.3 would be 1.7.3-asm.3. For the remainder of this section, "1.7.3-asm.3"
+will be used as a generic stand-in wherever an ASM version would be needed.
+"1.7" will be used as a stand-in wherever the minor version would be needed.
+
+## Processes in detail
+
+There are two paths for a commit to make it into a release branch. One is that
+a commit can land in master, and at some cadence will get rolled into
+staging-1.7-asm, then release-1.7-asm. The other path is for commits that are
+only relevant for a specific version, they can be merged directly into
+staging-1.7-asm and then into release-1.7-asm. There are no situations where
+any commits go from a release branch to any other branch, or from staging into
+master.
+
+The intention of having a staging branch at all is to allow for testing. There
+is no guarantee in the master branch that all of the components support the
+same versions at the same time. Having a staging branch ensures that the exact
+package of configuration that will be rolled into release gets tested together.
+
+The staging branch for the relevant version of ASM is created when the first RC
+version of ASM is released internally for testing. The branch should be created
+from the HEAD of master at the time, unless there is a known bad commit.
+
+In order to reduce confusion for anyone using the repo, the release-1.7-asm
+branch will not be created until ASM 1.7 is publicly released. Any development
+and testing until then will happen on staging-1.7-asm, or a specialized branch
+intended to be merged into staging at some point.
+
+## Motivation for the tag format
+
+This scheme has several desirable properties. Using tags instead of a commit
+hash is more user friendly, and the format of the tag clearly associates it
+with an ASM release so it will make mistakes more difficult to make. It's also
+easy for a human to compare two versions of the configuration and see which one
+is newer. Current methods of sorting still work with the new scheme (split on
+hyphen, semver the first element and lexical sort the second). It's well-formed
+enough to be verified mechanically.
+
+## Example release
+
+Releasing a configuration revision will follow this process:
+
+* A pull request is created with one or more commits from staging-1.7-asm
+* Review, tests, and any discussion happens on the PR
+* The PR gets merged into the release branch
+* An administrator pulls the release branch locally
+* The administrator uses a signed, annotated git tag to tag the commit
+* (Alternatively they can use GitHub's feature for verified signatures)
+* If this is a new ASM release, the tag changes to that release and N resets to
+1
+* Otherwise, the same tag as previous release is used except increment N by one
+* The administrator pushes the tag to GitHub
+
+## Diagram
+
+![Diagram of the process](./diagram.png?raw=true)
```

Comment:
minor: maybe we can update the right side of this diagram to be something more like a git tree.

Reply:
If someone better than me at diagrams wants to attempt that then I'd approve it immediately.